### PR TITLE
Auto Save (Fast) Mode

### PR DESCRIPTION
**Pull request by signedav**
fastmode

### How it works

Auto save (fast) mode can be activated in the settings.

In this mode it saves the feature:
- on every attribute value change (except the ones used for initialization after `resetAttribute()`).
- on every changed, added or removed vertex.

There are no buttons in the form. `confirm()` is still called when the form is closed by back key or tipping on the map to fire the needed close signals. Or - in case the constraints are not valid - cancel is called what means that nothing is saved.

Snapping and default values are set before the form opens, because they need to be saved on feature create (what happens when the geometry and constraints first gets valid).

### Constraints
A geometry is not saved when the constraints are not fulfilled.
A feature is saved as soon as the constraints are fulfilled for the first time.
When the constraints get unfulfilled *again* (means they have been fulfilled once, and the feature has been saved then), the feature is saved in the last valid state.

### Geometry
On digitizing the feature is saved as soon the geometry gets valid (and the constraints are fulfilled). When the geometry gets invalid *again* (by removing vertexes) the feature is deleted again. 
On edit geometry the behavior is different. There you cannot "invalidate" the geometry again. You cannot remove the last needed vertexes (3 on polygon / 2 on lines /  1 on point).

## To do, to make it fully consistent
- [x] Save feature geometry on vertex change (geometry editing)
- [x] Integrate confirm on close popup (by back key) in RelationCombobox
- [x] check with topological editing
- [x] do the optionals or deactivate auto save on these features
- [x] **fix new introduced issue in RelationCombobox**

## To do optionally
- [x] Integrate confirm on close popup (by back key) Tracking
- [x] Save features on fill ring - we do not save the geometry when valid because it's not decided if it should be a new feature or a ring hole

## Additional thoughts
Actually I don't really like the way we do it by setting featureCreated of the FeatureForm QML Object and make it dependant on that if we will create the feature or save it after we changed something. I would prefer something like it checks the feature of it's existance. But the issue here is, that the featureModel still has the "old" feature while creating (but not yet saved) the new one. This is for things like remeberings etc. Maybe I find a better way for that.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b57ea51e559f4e31f463364397168f278ecccf26-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b57ea51e559f4e31f463364397168f278ecccf26-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b57ea51e559f4e31f463364397168f278ecccf26-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b57ea51e559f4e31f463364397168f278ecccf26-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cd4f2c5bcaec110bb5cdd7c31a2e728b8a86cfc8-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cd4f2c5bcaec110bb5cdd7c31a2e728b8a86cfc8-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cd4f2c5bcaec110bb5cdd7c31a2e728b8a86cfc8-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cd4f2c5bcaec110bb5cdd7c31a2e728b8a86cfc8-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-952afb5b0d6f1ca534b74ca1ecd1b1bb7c342366-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-952afb5b0d6f1ca534b74ca1ecd1b1bb7c342366-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-952afb5b0d6f1ca534b74ca1ecd1b1bb7c342366-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-952afb5b0d6f1ca534b74ca1ecd1b1bb7c342366-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-1cc2b07db757af16059a6a6a19281d999c9b7f62-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-1cc2b07db757af16059a6a6a19281d999c9b7f62-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-1cc2b07db757af16059a6a6a19281d999c9b7f62-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-1cc2b07db757af16059a6a6a19281d999c9b7f62-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-0103ec6c270bdc03dc39a87a8398f42be0006832-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-0103ec6c270bdc03dc39a87a8398f42be0006832-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-0103ec6c270bdc03dc39a87a8398f42be0006832-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-0103ec6c270bdc03dc39a87a8398f42be0006832-x86_64.apk)

File: src/qml/RelationCombobox.qml [hunk 1]
```diff
@@ -134,59 +135,19 @@ Item {
...
+              var referencedValue = addFeaturePopup.attributeFormModel.attribute(relationCombobox._relation.resolveReferencedField(field.name))
+              comboBox.currentIndex = featureListModel.findKey(referencedValue)
+          }
       }
 
-      onClosed: {
-        addFeatureForm.active = false
+      onFeatureSaved: {
+          var referencedValue = addFeaturePopup.attributeFormModel.attribute(relationCombobox._relation.resolveReferencedField(field.name))
+          comboBox.currentIndex = featureListModel.findKey(referencedValue)
```
Comment (signedav) [inline src/qml/RelationCombobox.qml:150]: Here we have an issue on *not* auto save mode. This is happening before the featureListModel is reloaded (with timer) what means that it cannot find the value...

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-ee771c5b2b34f974ffa5f13456c6edecaad220b0-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-ee771c5b2b34f974ffa5f13456c6edecaad220b0-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-ee771c5b2b34f974ffa5f13456c6edecaad220b0-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-ee771c5b2b34f974ffa5f13456c6edecaad220b0-x86_64.apk)

**Comment by nirvn** [comment]
@signedav , since this PR and the one on topology overlap in areas of code, what should we do in terms of commit sequence? :)

**Comment by signedav** [comment]
Thanks @nirvn 

When #937 and #939 is ready soon, I suggest to merge it first and then I rebase this.

Btw. you want to do a review on this one?

Only missing thing is handling the tracking and the fill ring and I'm anyway not sure if auto save should be supported there at all.

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-6fbf55966b492d50054d2ffea8a1a3dc3c7bc7f2-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-6fbf55966b492d50054d2ffea8a1a3dc3c7bc7f2-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-6fbf55966b492d50054d2ffea8a1a3dc3c7bc7f2-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-6fbf55966b492d50054d2ffea8a1a3dc3c7bc7f2-x86_64.apk)

**Comment by nirvn** [comment]
@signedav , will review this afternoon.

File: src/qml/QFieldSettings.qml [hunk 2]
```diff
@@ -273,6 +275,45 @@ Page {
...
+                    Layout.fillWidth: true
+                    Layout.fillHeight: true
+                    Layout.minimumHeight: contentHeight
+                }
+
+            }
+
+            QfSwitch {
+                id: autoSaveCheckBox
+                checked: true
```
Comment (nirvn) [inline src/qml/QFieldSettings.qml:313]: The intent is for this new mode to be on by default?

File: src/qml/geometry_editors/VertexEditorToolbar.qml [hunk 3]
```diff
@@ -27,11 +27,18 @@ VisibilityFadingRow {
     featureModel.vertexModel.reset()
   }
 
+  function applyChanges( apply ) {
+    if ( apply && featureModel.vertexModel.dirty ){
+      featureModel.applyVertexModelToGeometry()
+      featureModel.save()
```
Comment (nirvn) [inline src/qml/geometry_editors/VertexEditorToolbar.qml:33]: OK so here, you'll have to re-create the vertex model based on modified feature geometry. It's needed due to the way we (now) keep track of new vs modified vertices in the context of topological editing.

File: src/qml/qgismobileapp.qml [hunk 4]
```diff
@@ -641,14 +641,61 @@ ApplicationWindow {
...
+                if( overlayFeatureFormDrawer.featureForm.featureCreated ) {
+                  //delete the feature when the geometry gets invalid again
+                  digitizingFeature.deleteFeature()
+                  overlayFeatureFormDrawer.featureForm.featureCreated = false
+                }
+            }
+        }
+      }
+
+      /*console.log
```
Comment (nirvn) [inline src/qml/qgismobileapp.qml:690]: To remove prior to merging?

File: src/qml/QFieldSettings.qml (hunk 2 above)
Comment (signedav) [reply src/qml/QFieldSettings.qml:313]: No it should be false.

File: src/qml/geometry_editors/VertexEditorToolbar.qml (hunk 3 above)
Comment (signedav) [reply src/qml/geometry_editors/VertexEditorToolbar.qml:33]: Okay, I'll check. Thanks.

**Comment by nirvn** [comment]
instant editing mode

reactive editing mode

...

**Comment by signedav** [comment]
I agree that auto save is not a good name for the function. I'm in favor of "fast editing mode". But I'm fine with your other suggestions as well.

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-a3f5eb9f9d5a52cc5bd8afdd64f1d1c7046f35e9-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-a3f5eb9f9d5a52cc5bd8afdd64f1d1c7046f35e9-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-a3f5eb9f9d5a52cc5bd8afdd64f1d1c7046f35e9-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-a3f5eb9f9d5a52cc5bd8afdd64f1d1c7046f35e9-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-095e0d810a4f7cce2185645a2df5d4a521db2e4b-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-095e0d810a4f7cce2185645a2df5d4a521db2e4b-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-095e0d810a4f7cce2185645a2df5d4a521db2e4b-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-095e0d810a4f7cce2185645a2df5d4a521db2e4b-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-92916d38d4a908fd1f922e1833dc00ac7ce28a86-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-92916d38d4a908fd1f922e1833dc00ac7ce28a86-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-92916d38d4a908fd1f922e1833dc00ac7ce28a86-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-92916d38d4a908fd1f922e1833dc00ac7ce28a86-x86_64.apk)

**Comment by signedav** [comment]
Hi @nirvn 
Can you have a look again. I set the original geometry with the current one on saving the feature. This should work with the topological editing.

Actually it's all done except an issue https://github.com/opengisch/QField/pull/936#pullrequestreview-389784183 what is not directly related but I plan to fix.

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-fcda2ffdbd08790082bb5db14c77319056766fd3-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-fcda2ffdbd08790082bb5db14c77319056766fd3-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-fcda2ffdbd08790082bb5db14c77319056766fd3-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-fcda2ffdbd08790082bb5db14c77319056766fd3-x86_64.apk)

**Comment by nirvn** [comment]
@signedav , I'll run a build locally and test things out.

**Comment by nirvn** [comment]
@signedav , something's not quire right. Moving a first node that's shared with other node(s) from oher feature(s) successfully moves those other nodes. But subsequent moves fail to move other overlapping nodes.

**Comment by nirvn** [comment]
@signedav , well, I just tried again and it's working. It could be that my first node was close to but not shared with a neighboring feature's node.

So, all good then.

**Comment by nirvn** [comment]
@signedav , it's quite nice to see the nodes from different features moving as you go along the nodes with the editor, nice one.

That being said, I found the UX a bit confusing here. The node movement is registered only when you change the currently selected node. Naturally, we can't handle it any other way. But, it's super confusing, I moved my node, then was just waiting to see something happen and nothing did.

I think for the node editor, you'll have to retain the (check) button to allow for a moved node to be saved without the need for the node selection to change.

**Comment by signedav** [comment]
Thanks a lot @nirvn

> I think for the node editor, you'll have to retain the (check) button to allow for a moved node to be saved without the need for the node selection to change.

Hm, I agree that it can be confusing.
So with the check the current vertex should be stored but the editing is not closed, right?

That would be nice, but don't you think it could be confusing as well then, because the user expects that he closes (and confirms) the editing with this?
Or should we close (and confirm) with the check? But then check and cancel (X) do not have any differences in the behavior.
So maybe I prefer having the check but do not close editing with it...

**Comment by nirvn** [comment]
@signedav , my preference align with yours, one check, no cancel. When user clicks on check, it saves relocation of node and keep editing on.

**Comment by signedav** [comment]
I reactivated the check to confirm the currently moved vertex, but I keep it grey like the other control buttons (instead of green on non-fast-mode). 

Looks like this, when pressing the check after moving a vertex.

Normal mode:


Fast mode:

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cf45d2c71f0e893554171d1cf5033396d93ce632-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cf45d2c71f0e893554171d1cf5033396d93ce632-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cf45d2c71f0e893554171d1cf5033396d93ce632-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-cf45d2c71f0e893554171d1cf5033396d93ce632-x86_64.apk)

**Comment by nirvn** [comment]
Nice. Good to see topological node movement too.

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-68f72b985307d3269a69264989c7cc4c4f7e8928-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-68f72b985307d3269a69264989c7cc4c4f7e8928-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-68f72b985307d3269a69264989c7cc4c4f7e8928-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-68f72b985307d3269a69264989c7cc4c4f7e8928-x86_64.apk)

**Comment by m-kuhn** [comment]
@nirvn @signedav 
What are the open questions here? Anything big?

**Comment by signedav** [comment]
@m-kuhn There is nothing open still. I did the implementation needed for the Relation Comobox and rebased the master. 
@nirvn Could you have a final look and approve if you approve :-)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b8069199c272e695918520659aceafc13549f4f2-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b8069199c272e695918520659aceafc13549f4f2-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b8069199c272e695918520659aceafc13549f4f2-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-b8069199c272e695918520659aceafc13549f4f2-x86_64.apk)

File: src/core/vertexmodel.h [hunk 5]
```diff
@@ -135,6 +135,13 @@ class VertexModel : public QStandardItemModel
 
     Q_INVOKABLE void removeCurrentVertex();
 
+
+    /**
+     * sets the geometry to the given \a geometry but preserves the index of the current vertex
+     * this is used to update the original geometry while still editing the model
```
Comment (m-kuhn) [inline src/core/vertexmodel.h:141]:
```suggestion
     * this is used to update the original geometry while still editing the model
     * \see `geometry` property
```

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-e3abb62203aa49f47d227d36f8bbf967312b1658-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-e3abb62203aa49f47d227d36f8bbf967312b1658-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-e3abb62203aa49f47d227d36f8bbf967312b1658-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-936-e3abb62203aa49f47d227d36f8bbf967312b1658-x86_64.apk)

**Comment by signedav** [comment]
Thanks @m-kuhn and @nirvn

**Pull request merged**